### DISCUSSION TITLE
Relevant executioner params for inelastic materials problems

### DISCUSSION THREAD

**Discussion by AndrewFalkowski** [idaholab/moose#21973] [Q&A Modules: Solid mechanics] [answered]
In trying to incorporate inelastic stress responses in my models, I've run into a number issues with convergence stability. The example input files included under the material docs for inelastic stress computation tend to have more sophisticated executioner blocks ([example](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/combined_creep_plasticity/combined_creep_plasticity.i)). I am wondering if there is any information on relevant executioner params (or considerations) for improving convergence in solving inelastic stress problems with creep and/or plastic deformation? Thanks!

**Comment by GiudGiud** [comment]
@hugary1995 @dschwen

**Comment by hugary1995** [comment, marked as answer]
I wonder if we don't already have a page talking about solvers in the tensor mechanics module. Anyways, here are some general considerations:
1. The non AD kernels/materials in the old system give inaccurate Jacobian, hence you are better off using PJFNK.
2. The AD kernels/materials give exact Jacobian, so Newton should perform better.
3. For preconditioner, a direct solver will generally perform the best if you have enough memory. If memory is a concern, you should use an iterative solver. Hyper/boomeramg is a very effective iterative solver for plasticity/creep problems.

Let us know which route you want to take, and we can provide more details regarding settings and solver parameters.

**Reply by AndrewFalkowski** [reply]
Thank you this was helpful. Assuming I am using the AD kerneIs, what would be the best approach for modeling plasticity and creep in a multi-material composite subject to thermal strains with materials exhibiting temperature dependent properties? And would you use different solver setup if you were solving plasticity or creep alone?

Also, could you elaborate a little more or point me to a resource discussing direct solver vs. iterative solver options?

Thanks

**Reply by hugary1995** [reply]
A direct solver can be specified by
```python
petsc_option_iname = '-pc_type'
petsc_option_value = 'lu'
```
Then choose your favorite parallel implementation of the matrix factorization package using `-mat_factor_solver_package`, typically your options are `superlu_dist`, `mumps` and `strumpack`.

Hypre/Boomeramg is documented here https://mooseframework.inl.gov/application_development/hypre.html

> Assuming I am using the AD kerneIs, what would be the best approach for modeling plasticity and creep in a multi-material composite subject to thermal strains with materials exhibiting temperature dependent properties? And would you use different solver setup if you were solving plasticity or creep alone?

This depends on the material. If the solid softens with temperature increase, then the nature of the problem won't change much compared to plasticity alone. Hence you can use similar options for the iterative solver.

If the solid hardens with temperature increase, then you'll be dealing with a potentially non-convex system. The usual path forward is
- lagging the temperature, or
- using fixed point iteration, or
- using a solver that handles local nonconvexity, e.g. trust-region or CG.

**Reply by hugary1995** [reply]
Well, actually in most cases there is heat loss due to thermal softening, and you are back to an easy, convex problem, hence you probably don't need the "usual path forward" methods...

**Reply by AndrewFalkowski** [reply]
Ah ok that makes a lot of sense. I seem to be getting a more stable solve, but I think my tolerances are off as I am having trouble with nonlinear convergence. How (and under what block) should one go about setting these tolerances? I was reading through this [page](https://mooseframework.inl.gov/modules/tensor_mechanics/Convergence.html), but I'm still not sure about the general application of these.

**Reply by AndrewFalkowski** [reply]
Realized this might be because I am near a steady state, but general advice on setting those tolerances would be appreciated.
